src/dyad/modules/dyad.c

```diff
@@ -182,125 +213,205 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
     rc = ctx->dtl_handle->send (ctx->dtl_handle, inbuf, inlen);
     FLUX_LOG_INFO (h, "Close DTL connection with consumer");
     ctx->dtl_handle->close_connection (ctx->dtl_handle);
-    free (inbuf);
+    ctx->dtl_handle->return_buffer (ctx->dtl_handle, &inbuf);
     if (DYAD_IS_ERROR (rc)) {
         FLUX_LOG_ERR (ctx->h, "Could not send data to client via DTL\n");
         errno = ECOMM;
         goto fetch_error;
     }
 
     FLUX_LOG_INFO (h, "Close RPC message stream with an ENODATA (%d) message", ENODATA);
+    DYAD_PERF_REGION_BEGIN (ctx->perf_handle, "dyad_server_send_response");
     if (flux_respond_error (h, msg, ENODATA, NULL) < 0) {
-        FLUX_LOG_ERR (h,
-                      "DYAD_MOD: %s: flux_respond_error with ENODATA failed\n",
-                      __FUNCTION__);
+        FLUX_LOG_ERR (h, "DYAD_MOD: %s: flux_respond_error with ENODATA failed\n", __FUNCTION__);
     }
+    DYAD_PERF_REGION_END (ctx->perf_handle, "dyad_server_send_response");
     errno = saved_errno;
     FLUX_LOG_INFO (h, "Finished dyad.fetch module invocation\n");
-    return;
+    goto end_fetch_cb;
 
 fetch_error:
     FLUX_LOG_ERR (h, "Close RPC message stream with an error (errno = %d)\n", errno);
+    DYAD_PERF_REGION_BEGIN (ctx->perf_handle, "dyad_server_send_response");
     if (flux_respond_error (h, msg, errno, NULL) < 0) {
         FLUX_LOG_ERR (h, "DYAD_MOD: %s: flux_respond_error", __FUNCTION__);
     }
+    DYAD_PERF_REGION_END (ctx->perf_handle, "dyad_server_send_response");
     errno = saved_errno;
     return;
+
+end_fetch_cb:
+    DYAD_PERF_REGION_END (ctx->perf_handle, "dyad_server_cb");
+    return;
 }
 
-static dyad_rc_t dyad_open (flux_t *h, dyad_dtl_mode_t dtl_mode, bool debug)
+static dyad_rc_t dyad_open (flux_t *h, dyad_dtl_mode_t dtl_mode, bool debug, optparse_t *opts)
 {
     dyad_mod_ctx_t *ctx = getctx (h);
     dyad_rc_t rc = 0;
     char *e = NULL;
 
     ctx->debug = debug;
-    rc = dyad_dtl_init (&(ctx->dtl_handle), dtl_mode, h, ctx->debug);
-
+    rc = dyad_perf_init (&(ctx->perf_handle), true, opts);
+    if (DYAD_IS_ERROR (rc))
+        goto open_done;
+    rc = dyad_dtl_init (&(ctx->dtl_handle),
+                        dtl_mode,
+                        DYAD_COMM_SEND,
+                        h,
+                        ctx->debug,
+                        ctx->perf_handle);
```

Comment:
This is relevant to DTL except the last argument. The only thing added is the third argument, DYAD_COMM_SEND.